### PR TITLE
GFX & Shader Documentation

### PR DESCRIPTION
**Pull request by guusw**
## [Preview](https://tdrz.nl/shards_docs/)

Document GFX shards:
- [x] BuiltinFeature
- [x] Draw
- [x] Drawable
- [x] DrawablePass
- [x] DrawQueue
- [x] Feature
- [x] glTF
- [x] Render
- [x] UIPass

### CODE REVIEW

**Comment by guusw** [comment]
Opening it now for review since it might get too big. Since this also contains a bunch of technical fixes/changes.

I still plan to continue documenting the rest of the graphics shards, add more samples & add more details on some topics.

File: docs/details/shards/GFX/BuiltinFeature.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+This shard generates features with preset behavior, the behavior depends on the [Id](../../../enums/BuiltinFeatureId)
+
+## Built-in Feature Ids
+
+### Transform
+
+This feature implements basic world/view/projection transform.
+
+Here are the entry points that is defines and can be added as dependencies by your own features.
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/BuiltinFeature.md:9]:
```suggestion
Here are the entry points that are defined and can be added as dependencies by your own features.
```

File: docs/details/shards/GFX/BuiltinFeature.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+Transforms the object normal and writes the transformed result into the `worldNormal` global (float3).
+If the mesh doesn't have vertex normals the normal will be `(0.0, 0.0, 1.0)`
+
+#### `writeNormal` (vertex)
+
+Writes `worldNormal` global into `worldNormal` output
+
+### BaseColor
+
+This feature adds a per-object shader parameter with the name `baseColor` and texture parameter with the same name
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/BuiltinFeature.md:36]:
```suggestion
This feature adds a per-object shader parameter with the name `baseColor` and a texture parameter with the same name
```

File: docs/details/shards/GFX/BuiltinFeature.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+Here are the entry points that is defines and can be added as dependencies by your own features.
+
+#### `initColor` (vertex)
+
+This vertex shader entry point sets up the the `color`  global (float4) with the vertex color from the mesh if set. If the mesh doesn't have a vertex color it'll be white with alpha 1.0.
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/BuiltinFeature.md:42]:
```suggestion
This vertex shader entry point sets up the `color`  global (float4) with the vertex color from the mesh if it has been set. If the mesh doesn't have a vertex color, it'll be white with alpha 1.0.
```

File: docs/details/shards/GFX/BuiltinFeature.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+#### VertexColorFromNormal
+
+This feature outputs the mesh normal as a per-vertex color output.
+
+Use for debugging
+
+### Wireframe
+
+This feature changes how objects are rendered so that their edges are highlighted. Very basic, fixed edge width & color.
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/BuiltinFeature.md:68]:
```suggestion
This feature changes how objects are rendered so that their edges are highlighted. Very basic, with a fixed edge width and color.
```

File: docs/details/shards/GFX/BuiltinFeature.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+### Transform
+
+This feature implements basic world/view/projection transform.
+
+Here are the entry points that is defines and can be added as dependencies by your own features.
+
+#### `initLocalPosition` (vertex)
+
+Sets the `localPosition` (float4) global from the vertex position
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/BuiltinFeature.md:13]:
```suggestion
Sets the `localPosition` (float4) global from the vertex position.
```

File: docs/details/shards/GFX/BuiltinFeature.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+Sets the `localPosition` (float4) global from the vertex position
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/BuiltinFeature.md:15]: Duplicate line.

File: docs/details/shards/GFX/BuiltinFeature.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+#### `initScreenPosition` (vertex)
+
+Sets the `screenPosition` global (float4) to the transformed vertex position (`=proj*view*world*<vertex>`)
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/BuiltinFeature.md:19]:
```suggestion
Sets the `screenPosition` global (float4) to the transformed vertex position (`=proj*view*world*<vertex>`).
```

We should keep the punctuation consistent for our docs - a period is required at the end of every sentence.

File: docs/details/shards/GFX/BuiltinFeature.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+Writes `screenPosition` global to `position` output
+
+#### `initWorldNormal` (vertex)
+
+Transforms the object normal and writes the transformed result into the `worldNormal` global (float3).
+If the mesh doesn't have vertex normals the normal will be `(0.0, 0.0, 1.0)`
+
+#### `writeNormal` (vertex)
+
+Writes `worldNormal` global into `worldNormal` output
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/BuiltinFeature.md:32]:
```suggestion
Writes `worldNormal` global into `worldNormal` output.
```

File: docs/details/shards/GFX/BuiltinFeature.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+This vertex shader entry point sets up the the `color`  global (float4) with the vertex color from the mesh if set. If the mesh doesn't have a vertex color it'll be white with alpha 1.0.
+
+#### `writeColor` (vertex)
+
+Writes `color` global to `color` output
+
+#### `readColor` (fragment)
+
+Reads `color` input into `color` global (float4)
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/BuiltinFeature.md:50]:
```suggestion
Reads `color` input into `color` global (float4).
```

File: docs/details/shards/GFX/BuiltinFeature.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+#### `textureColor` (fragment)
+
+Reads the texture color from the `baseColor` texture and multiplies it with the current `color` global (if the texture exists, otherwise it does nothing)
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/BuiltinFeature.md:54]:
```suggestion
Reads the texture color from the `baseColor` texture and multiplies it with the current `color` global. It does nothing if the texture does not exist.
```

File: docs/details/shards/GFX/BuiltinFeature.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+### Wireframe
+
+This feature changes how objects are rendered so that their edges are highlighted. Very basic, fixed edge width & color.
+
+Use for debugging
+
+
+### Velocity
+
+This feature outputs per-object velocity into a `velocity` output and `velocity` global. For usage with effects that require a velocity buffer such as motion blur or temporal anti-aliasing
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/BuiltinFeature.md:75]:
```suggestion
This feature outputs per-object velocity into a `velocity` output and `velocity` global. For usage with effects that require a velocity buffer such as motion blur or temporal anti-aliasing.
```

File: docs/details/shards/GFX/Draw.md [hunk 12]
```diff
@@ -0,0 +1,7 @@
+Draw takes a [Drawable](../Drawable) and adds a render command for it to a [RenderQueue](../RenderQueue)
+
+--8<-- "includes/drawable_refactor.md"
+
+## Default queue
+
+If no render queue is specified, it goes to a global default queue
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/Draw.md:7]:
```suggestion
If no render queue is specified, it goes to a global default queue.
```

File: docs/details/shards/GFX/Drawable.md [hunk 13]
```diff
@@ -0,0 +1,47 @@
+The drawable shard is used to define an instance of a drawn object, it usually has a transform (in world space) and a mesh to draw.
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/Drawable.md:1]:
```suggestion
The drawable shard defines an instance of a drawn object. It usually has a transform (in world space) and a mesh to draw.
```

File: docs/details/shards/GFX/Drawable.md [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+--8<-- "includes/drawable_refactor.md"
+
+## Params
+
+The `Params` parameter contains a table of values that will be passed to the shader when this drawable is rendered.
+
+!!! info Defining shader parameters
+    See [GFX.Feature](../Feature) about where shader parameters are defined and used
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/Drawable.md:10]:
```suggestion
    See [GFX.Feature](../Feature) about where shader parameters are defined and used.
```

File: docs/details/shards/GFX/Drawable.md [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+Constant parameters are passed through the input table. They are read and stay the same until this function is called again.
+
+```clojure
+{:Mesh .mesh :Transform .transform} (GFX.Drawable) >= .my-drawable
+```
+
+### Dynamic parameters
+
+Dynamic parameters are passed as parameters to the shard, the same fields can be set, except for the mesh. They are referenced and when their value changes will reflect this change in the rendered result.
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/Drawable.md:26]:
```suggestion
Dynamic parameters are passed in as parameters to the shard. The same fields can be set elsewhere, with the exception of the mesh. When they are referenced, changes in their value will be reflected in the rendered result.
```

Might need to reword this as the definition is a little unclear.

File: docs/details/shards/GFX/Drawable.md [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+```clojure
+.. >= .dynamic-color
+.. >= .dynamic-transform
+{:Mesh .mesh} (Drawable :Transform .dynamic-transform :Params {:baseColor .dynamic-color}) >= .my-drawable
+```
+
+## Intended usage
+
+The currently intended way to set up drawables without frequent changes is to set them up from within a setup block. Any dynamic parameters that might change should then be passed as a parameter to the shard.
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/Drawable.md:36]:
```suggestion
To avoid frequent changes, you should set up drawables in a `Setup` block and pass their dynamic parameters to the shard.
```
Might need to reword this too as the definition is a little unclear.

File: docs/details/shards/GFX/DrawablePass.md [hunk 17]
```diff
@@ -0,0 +1,3 @@
+A step that can be passed to [Render](../Render). When processed it will take all drawables from the referenced [DrawQueue](../DrawQueue) and draw them.
+
+Features can be specified that will be applied to
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/DrawablePass.md:3]: Incomplete sentence.

File: docs/details/shards/GFX/Feature.md [hunk 18]
```diff
@@ -0,0 +1,13 @@
+A feature can be though of as an extension or plugin to the rendering logic.
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/Feature.md:1]:
```suggestion
A feature can be thought of as an extension or plugin to the rendering logic.
```

File: docs/details/shards/GFX/Feature.md [hunk 19]
```diff
@@ -0,0 +1,13 @@
...
+
+The may have various components:
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/Feature.md:3]:
```suggestion
It may have various components:
```

File: docs/details/shards/GFX/Feature.md [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+
+- Blend State (Alpha, Additive, ...)
+- Depth Testing
+- Vertex/Fragment shaders
+
+For a more extensive description about shaders see [Shaders](../Shader/index.md)
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/Feature.md:9]:
```suggestion
For a more extensive description about shaders, see [Shaders](../Shader/index.md).
```

File: docs/details/shards/GFX/Render.md [hunk 21]
```diff
@@ -0,0 +1,16 @@
+This function should be used only once per rendered frame.
+
+It takes a [View](../View) or sequence of [Views](../View) that represent the main camera as a parameter.
+
+??? info Default view
+    If neither `:View` nor `:Views` is specified the renderer will use a default view with both identity view and projection matrices.
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/Render.md:6]:
```suggestion
    If neither `:View` nor `:Views` is specified, the renderer will use a default view with both identity view and projection matrices.
```

File: docs/details/shards/GFX/Render.md [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+
+The `:Steps` parameter contains the list of rendering operations to perform, this can be sequence objects created by one of the following:
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/Render.md:8]:
```suggestion
The `:Steps` parameter contains the list of rendering operations to perform. This can be a sequence of objects created by one of the following:
```

File: docs/details/shards/GFX/Render.md [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+
+- [DrawablePass](../DrawablePass)
+- [EffectPass](../EffectPass)
+- [UIPass](../UIPass)
+
+## Default queue
+
+If no render queue is specified, this command reads from a global default queue
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/Render.md:16]:
```suggestion
If no render queue is specified, this command reads from a global default queue.
```

File: docs/details/shards/GFX/UIPass.md [hunk 24]
```diff
@@ -0,0 +1,3 @@
+This defines a render step for rendering [UI](../../General/UI)
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/UIPass.md:1]:
```suggestion
Defines a render step for rendering [UI](../../General/UI).
```

File: docs/details/shards/GFX/UIPass.md [hunk 25]
```diff
@@ -0,0 +1,3 @@
...
+
+Check the examples on some of the UI shards for how to use this step, for example [UI.Label](../../UI/Label)
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/UIPass.md:3]: Might be better to directly insert the examples here instead of linking it elsewhere for usage examples.

File: docs/details/shards/GFX/glTF.md [hunk 26]
```diff
@@ -0,0 +1,41 @@
+Loads a glTF model from a file. Outputs a `GFX.DrawableHierarchy` that can be passed to [Draw](../Draw)
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/glTF.md:1]:
```suggestion
Loads a glTF model from a file. Outputs a `GFX.DrawableHierarchy` that can be passed to [Draw](../Draw).
```

File: docs/details/shards/GFX/glTF.md [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+
+--8<-- "includes/drawable_refactor.md"
+
+## Static file
+
+Loaded when the wire containing it is warmed up
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/glTF.md:7]:
```suggestion
Loaded when the Wire containing it is warmed up.
```

File: docs/details/shards/GFX/glTF.md [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+
+```clojure
+{:Transform ...} (GFX.glTF :Path "pathToModel.glb") >= .drawable
+```
+
+## Dynamic file
+
+Loaded when activated.
+You should cache results inside `(Setup ...)` shard
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/glTF.md:16]:
```suggestion
You should cache the results inside a `(Setup ...)` shard.
```

File: docs/details/shards/GFX/glTF.md [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+
+```clojure
+{:Transform ... :Path .path} (GFX.glTF) >= .drawable
+```
+
+## Raw bytes
+
+Loaded when activated.
+You should cache results inside `(Setup ...)` shard
```
Comment (ChelseaChX) [inline docs/details/shards/GFX/glTF.md:25]:
```suggestion
You should cache the results inside a `(Setup ...)` shard.
```

File: docs/docs/reference/shards/GFX/index.md [hunk 30]
```diff
@@ -0,0 +1,5 @@
+# GFX
+
+Here you'll find all the shards related to graphics and rendering.
+
+For a basic example, check the [GFX.Render](./Render) shard documentation
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/GFX/index.md:5]:
```suggestion
For a basic example, check out the [GFX.Render](./Render) shard documentation.
```

File: docs/docs/reference/shards/Shader/index.md [hunk 31]
```diff
@@ -0,0 +1,185 @@
+# Shaders
+
+Shaders are written using a subset of the regular shards language, in addition to some of the shader specific shards listed here
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/Shader/index.md:3]:
```suggestion
Shaders are written using a subset of the regular shards language, in addition to some of the shader specific shards listed here.
```

File: docs/docs/reference/shards/Shader/index.md [hunk 32]
```diff
@@ -0,0 +1,185 @@
...
+```clojure
+.v3-forward
+(| (Take 0) >= .x)
+(| (Take 1) >= .y)
+(| (Take 2) >= .z)
+```
+
+### Push (for matrix types only)
+
+Push works for building matrix types out of vectors, but with the constraint that it they can't be used from within branching control flows.
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/Shader/index.md:41]:
```suggestion
`Push` works for building matrix types out of vectors, but with the constraint that the matrices can't be used from within branching control flows.
```

File: docs/docs/reference/shards/Shader/index.md [hunk 33]
```diff
@@ -0,0 +1,185 @@
...
+```clojure
+(ForRange 0 32 (->
+  = .i
+  ...
+  ))
+```
+
+### Function-like wire evaluation
+
+Pure wires are supported, they can be ran by using do and passing an input as a scalar/vector/matrix type. Their output is returned
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/Shader/index.md:150]:
```suggestion
Pure Wires are supported. They can be run using `Do` and passing an input as a scalar/vector/matrix type. Their output is returned.
```

File: docs/docs/reference/shards/Shader/index.md [hunk 34]
```diff
@@ -0,0 +1,185 @@
...
+  (Float2 .neg-s .c) >> .result
+  .result ; Output a 2x2 matrix
+)
+
+0.45 (Do rotation-matrix) >= .mat2-2 ; This will contain the result from the wire
+
+```
+
+Non-pure wires are also supported they will also accept an input value and output their output.
+In addition they will also capture any variables referenced that are defined in parent wires
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/Shader/index.md:169]:
```suggestion
In addition they will also capture any variables referenced that are defined in parent Wires.
```

File: docs/docs/reference/shards/Shader/index.md [hunk 35]
```diff
@@ -0,0 +1,185 @@
...
+
+3.0 >== .global
+(Do sub-logic) >= .v1
+2.0 > .global
+(Do sub-logic) >= .v2
+```
+
+!!! info "Partial support"
+
+    Only values that are read from parent wires are currently supported, trying to update a variable defined in a parent wire will have no effect on that variable
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/Shader/index.md:185]:
```suggestion
    Currently, only values that are read from parent Wires are supported. Trying to update a variable defined in a parent Wire will have no effect on that variable.
```

File: docs/docs/reference/shards/UI/index.md [hunk 36]
```diff
@@ -0,0 +1,9 @@
+# UI
+
+Shards for rendering user interfaces.
+
+!!! info
+
+    All the shards listed here should be placed within a [`UI`](../General/UI.md) shard
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/UI/index.md:7]:
```suggestion
    All the shards listed here should be placed within a [`UI`](../General/UI.md) shard.
```

File: docs/docs/reference/shards/index.md [hunk 37]
```diff
@@ -0,0 +1,76 @@
+---
+authors: Fragcolor & contributors
+license: CC-BY-SA-4.0
+---
+
+# Shards
+
+This section contains a reference of all the shards available within the shards language.
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/index.md:8]:
```suggestion
This section contains a reference of all the shards available within the Shards language.
```

File: docs/docs/reference/shards/index.md [hunk 38]
```diff
@@ -0,0 +1,76 @@
...
+
+## How to read
+
+To see how to read the shard description, let's look at [`String.Join`](./String/Join) as an example:
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/index.md:12]:
```suggestion
Let's look at the description for [`String.Join`](./String/Join) as an example:
```

File: docs/docs/reference/shards/index.md [hunk 39]
```diff
@@ -0,0 +1,76 @@
...
+| `Separator` |  | The string to use as a separator. | `""` | `String` |
+</div>
+
+The first thing you will see on every shard page is this table.
+
+The first two entries in the table describe what inputs a shard will accept and what kind of values it will output.
+
+### &lt;input&gt;
+
+In this case the `String.Join` shard accepts a sequence of `Strings` as an input, written as `[ String ]`, the square brackets meaning sequence.
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/index.md:28]:
```suggestion
In this case, the `String.Join` shard accepts a sequence of `Strings` as an input - written as `[String]`. Sequences are written with its data type enclosed by square brackes.
```

File: docs/docs/reference/shards/index.md [hunk 40]
```diff
@@ -0,0 +1,76 @@
...
+
+### &lt;output&gt;
+
+The output will be a single `String`
+
+### Parameters
+
+After the input and output of the shard, the parameters are listed. The `String.Join` shard requires a `Separator` parameter and it has to be a `String`, the default value when not specified will be `""`
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/index.md:36]:
```suggestion
After the input and output of the shard, the parameters are listed. The `String.Join` shard requires a `Separator` parameter that is a `String`. When the parameter is not specified, the default value will be `""`.
```

File: docs/docs/reference/shards/index.md [hunk 41]
```diff
@@ -0,0 +1,76 @@
...
+
+## Optional parameters
+
+<div class="sh-parameters" markdown="1">
+| Name | - {: #sh-flags-row} | Description | Default | Type |
+|------|---------------------|-------------|---------|------|
+| `Window` | :fontawesome-solid-circle-plus:{title="Optional"}  | None or a window variable we wish to use as relative origin. | `None` | `Object ` |
+</div>
+
+Occasionally you will also see a :fontawesome-solid-circle-plus: icon next to a parameter like this, this means that the parameter is optional and it's value can be left unset.
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/index.md:46]:
```suggestion
Occasionally you will see a :fontawesome-solid-circle-plus: icon next to a parameter. This means that the parameter is optional and its value can be left unset.
```

File: docs/docs/reference/shards/index.md [hunk 42]
```diff
@@ -0,0 +1,76 @@
...
+
+When a variable is not optional and it's default value is not `None` you should set the parameter on the shard.
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/index.md:48]:
```suggestion
When a variable is not optional and its default value is not `None`, you should set the parameter on the shard.
```

File: docs/docs/reference/shards/index.md [hunk 43]
```diff
@@ -0,0 +1,76 @@
...
+
+## Type descriptions
+
+The type column shows the expected types for a input/output/parameter
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/index.md:52]:
```suggestion
The type column shows the expected types for an input/output/parameter:
```

File: docs/docs/reference/shards/index.md [hunk 44]
```diff
@@ -0,0 +1,76 @@
...
+| `Values` | `String Float` |
+| `Sequence` | `[ String Float ]` |
+| `Table` | `{ String Float }` |
+| `Variables` | `&String &Float` |
+| `SequenceVariable` | `&[ Float4 ]` |
+</div>
+
+Multiple different types in a row indicate that any of the types listed are accepted.
+
+When placed within `[ square brackets ]` this means that a [Sequence](./types/#sequence) of those types is accepted
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/index.md:66]:
```suggestion
When placed within `[ square brackets ]` this means that a [Sequence](./types/#sequence) of those types is accepted.
```

File: docs/docs/reference/shards/index.md [hunk 45]
```diff
@@ -0,0 +1,76 @@
...
+
+When placed within `{ curly brackets }` this means that a [Table](./types/#table) of those types as values accepted
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/index.md:68]:
```suggestion
When placed within `{ curly brackets }` this means that a [Table](./types/#table) of those types as values is accepted.
```

File: docs/docs/reference/shards/index.md [hunk 46]
```diff
@@ -0,0 +1,76 @@
...
+
+When a value is prefixed with an `&ampersand` this means that a variable / [ContextVar](./types/#contextvar) of that type is accepted
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/index.md:70]:
```suggestion
When a value is prefixed with an `&ampersand`, this means that a variable / [ContextVar](./types/#contextvar) of that type is accepted.
```

File: docs/docs/reference/shards/index.md [hunk 47]
```diff
@@ -0,0 +1,76 @@
...
+
+## Further reading
+
+See the [Types](./types) page for more information about types
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/index.md:74]:
```suggestion
See the [Types](./types) page for more information about types.
```

File: docs/docs/reference/shards/types.md [hunk 48]
```diff
@@ -63,17 +63,17 @@ For example, **Any** as the allowed data type for input and `:Value` parameter o
...
     ```
 
-## Array
+## Sequence
 
-Type **Array** is a collection of values that can be accessed directly via indexes (since items are indexed by contiguous integers).
+Type **Sequence** is a collection of values that can be accessed directly via indexes (since items are indexed by contiguous integers).
 
-Also called a vector. Looks exactly like the [**Seq**](#seq) type, but an **Array** type's items are accessible by index (and hence accessible randomly). Example of an **Array** type would be: `[43 6 1]`.
+Also called a vector or an array. Looks exactly like the [**Seq**](#seq) type, but a **Sequence** type's items are accessible by index (and hence accessible randomly). Example of an **Sequence** type would be: `[43 6 1]`.
```
Comment (ChelseaChX) [inline docs/docs/reference/shards/types.md:70]:
```suggestion
It is also called a vector or an array. A **Sequence**'s items are accessible by index. Example of an **Sequence** type would be: `[43 6 1]`.
```

**Comment by ChelseaChX** [comment]
Looks good! Going forward we should be trying to standardize the writing of docs for Shards. `Wire` and `Mesh` are Shard specific terms and should have their first letter capitalized to distinct them from regular wires / meshes. We'll have to clean up on the punctuation too to make our docs seem less sloppy. Thanks for helping with the docs on the UI / GFX shards, as well as streamlining the shards doc pages to make it easier to peruse.

**Review by Kryptos-FR — CHANGES_REQUESTED** [review]
There are too many changes that are unrelated with each others. Some are based on misconception on how the documentation is generated (like the changes related to "verbose" logs which is not needed since the samples are run in release mode).

The list of samples in generated doc is broken.

help ()shouldn't have any markdown or HTML  (or even line return) in it as it will be used by the visual scripting and it is undecided at this stage if we will support markdown there (or another syntax). That's why we have the Details section, where more "complex" documentation can be manually written.

Please split into several PRs, starting with all the changes related to GFX code being separate, then every new doc feature being also separate so we can select the one we want.

- PR related to search
- PR related to reformatting the input/output/parameter into a single table (for which I don't see why we need HTML, as markdown is capable of doing tables)

Also at this stage, I wouldn't move any pages around since this is still a work in progress. That should be part of another PR and ongoing discussion.

File: docs/docs/reference/shards/index.md (hunk 37 above)
Comment (guusw) [reply docs/docs/reference/shards/index.md:8]: so we user "Shards" (with a capital) when referring to the language and "shards" when referring to the building blocks?

File: docs/docs/reference/shards/index.md (hunk 42 above)
Comment (guusw) [reply docs/docs/reference/shards/index.md:48]: 😅

File: docs/details/shards/GFX/Drawable.md (hunk 16 above)
Comment (guusw) [reply docs/details/shards/GFX/Drawable.md:36]: Reworded

File: docs/details/shards/GFX/Drawable.md (hunk 15 above)
Comment (guusw) [reply docs/details/shards/GFX/Drawable.md:26]: Reworded

File: docs/details/shards/GFX/Feature.md (hunk 19 above)
Comment (guusw) [reply docs/details/shards/GFX/Feature.md:3]: reworded

File: docs/docs/reference/shards/Shader/index.md (hunk 32 above)
Comment (guusw) [reply docs/docs/reference/shards/Shader/index.md:41]: Reworded

File: src/extra/gfx/feature.cpp [hunk 49]
```diff
@@ -81,22 +91,27 @@ struct BuiltinFeatureShard {
...
-      }
-      :DrawData [(-> ...)]/(-> ...)
-      :Params [
-        {:Name <string> :Type <type> :Dimension <number>}
-      ]
-    }
-  */
   static SHTypesInfo inputTypes() { return CoreInfo::AnyTableType; }
   static SHTypesInfo outputTypes() { return Types::Feature; }
+  static SHOptionalString help() {
```
Comment (sinkingsugar) [inline src/extra/gfx/feature.cpp:96]: There is a serious discussion here to be made about embedded docs using markdown or not. I think it's time to have it.
I'd say yes as it would help current generation. 
@Kryptos-FR also our current UI shards should easily support markdown right?

Comment (guusw) [reply src/extra/gfx/feature.cpp:96]: Yeah was thinking we might want to show some more elaborate help strings, I know this package does at least some basic markdown like code blocks: https://github.com/lampsitter/egui_commonmark

File: docs/mkdocs.yml [hunk 50]
```diff
@@ -23,6 +23,7 @@ extra:
       name: Company profile on LinkedIn
 markdown_extensions:
   - admonition
+  - attr_list
```
[the remark this answers is not in the document]
Comment (guusw) [reply docs/mkdocs.yml:26]: Yes, there's a table cell somewhere that has an attribute list currently

File: docs/mkdocs.yml [hunk 51]
```diff
@@ -52,9 +57,9 @@ theme:
   font:
     code: Fira Code
     text: Montserrat
-  language: en
+  language: custom
```
[the remark this answers is not in the document]
Comment (guusw) [reply docs/mkdocs.yml:60]: This was required for the search to set some properties, see [this](https://github.com/fragcolor-xyz/shards/blob/621844d6662c12842b30a5c004054651f9b1594a/docs/overrides/partials/languages/custom.html) file

The main thing of interest is `"search.config.pipeline": "stopWordFilter",` since it removes the other default filter that filters out symbols, so you couldn't search for ">=", ">>",etc. for example

Unless there's an other way to set this (couldn't figure it out), I'm setting it this way. If we ever decide to build other language documentation we can split this.

File: docs/docs/reference/shards/UI/index.md [hunk 52]
```diff
@@ -0,0 +1,9 @@
+# UI
```
[the remark this answers is not in the document]
Comment (guusw) [reply docs/docs/reference/shards/UI/index.md:1]: https://github.com/fragcolor-xyz/shards/pull/476#discussion_r1040947770

File: src/extra/gfx/feature.cpp (hunk 49 above)
Comment (sinkingsugar) [reply src/extra/gfx/feature.cpp:96]: let's implement https://github.com/lampsitter/egui_commonmark

**Comment by guusw** [comment]
Rebased & cleaned up

File: src/extra/gfx/feature.cpp [hunk 53]
```diff
@@ -15,27 +15,37 @@
...
+enum class BuiltinFeatureId { Transform, BaseColor, VertexColorFromNormal, Wireframe, Velocity };
+}
+
+ENUM_HELP(gfx::BuiltinFeatureId, gfx::BuiltinFeatureId::Transform, SHCCSTR("Add basic world/view/projection transform"));
+ENUM_HELP(gfx::BuiltinFeatureId, gfx::BuiltinFeatureId::BaseColor,
+          SHCCSTR("Add basic color from vertex color and (optional) color texture"));
+ENUM_HELP(gfx::BuiltinFeatureId, gfx::BuiltinFeatureId::VertexColorFromNormal, SHCCSTR("Outputs color from vertex color"));
+ENUM_HELP(gfx::BuiltinFeatureId, gfx::BuiltinFeatureId::Wireframe, SHCCSTR("Modifies the main color to visualize vertex edges"));
+ENUM_HELP(gfx::BuiltinFeatureId, gfx::BuiltinFeatureId::Velocity,
+          SHCCSTR("Outputs object velocity into the velocity global & output"));
```
Comment (Kryptos-FR) [inline src/extra/gfx/feature.cpp:28]: @ChelseaChX Should we have a final `.` for those help strings? I added one on my own enum description since we also had those on the regular `help()` function.

**Review by Kryptos-FR — APPROVED** [review]
LGTM.

Just a remark regarding documenting the enums.

**Review by sinkingsugar — APPROVED** [review]
Good as a starting point I guess. But needs way more love.

File: src/extra/gfx/feature.cpp (hunk 53 above)
Comment (guusw) [reply src/extra/gfx/feature.cpp:28]: Yeah I think we have them everywhere else in help strings, can add them here.

**Comment by guusw** [comment]
@sinkingsugar Yeah not complete yet, but I'll open a new PR for the remaining docs

File: src/extra/gfx/feature.cpp (hunk 53 above)
Comment (ChelseaChX) [reply src/extra/gfx/feature.cpp:28]: > @ChelseaChX Should we have a final `.` for those help strings? I added one on my own enum description since we also had those on the regular `help()` function.

Yeah we should keep it consistent.
